query/src/common/evaluator/eval_context.rs

```diff
@@ -0,0 +1,106 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common_datablocks::DataBlock;
+use common_datavalues::ColumnRef;
+use common_datavalues::DataField;
+use common_datavalues::DataTypeImpl;
+use common_exception::ErrorCode;
+use common_exception::Result;
+
+use crate::sql::IndexType;
+
+pub trait EvalContext {
+    type VectorID: PartialEq;
+
+    fn get_vector(&self, id: &Self::VectorID) -> Result<TypedVector>;
+
+    fn insert_vector(&mut self, id: Self::VectorID, vector: TypedVector) -> Result<()>;
+
+    fn tuple_count(&self) -> usize;
+}
+
+impl EvalContext for DataBlock {
+    type VectorID = IndexType;
+
+    fn get_vector(&self, id: &IndexType) -> Result<TypedVector> {
+        let name = id.to_string();
```

Comment:
IndexType is aliased to usize. Why we use it as name ?

Reply:
It's actually an unique ID of each column that can be referenced in a query. `ColumnIndex` is bound with `Metadata` by binder, so we can just utilize with it in executor.

A `EvalContext` is an unordered set of variables(i.e. `TypedVector`), each variable is assigned with a unique ID(`IndexType` for `ScalarExpr`). What we want here is a `HashMap<IndexType, TypedVector>` rather than a `DataBlock`, but we can keep it for now.

Reply:
And it's possible to use a string instead, but I prefer to use `IndexType` here.